detox/runners/jest/reporters/DetoxReporter.js

```diff
@@ -25,38 +29,99 @@ class DetoxReporter extends JestVerboseReporter {
     }
   }
 
+  onTestFileStart(test) {
+    this._pendingTestFiles.add(test.path);
+
+    // @ts-ignore Precaution in case Jest migrates to the new signature
+    if (typeof super.onTestFileStart === 'function') {
+      // @ts-ignore
+      super.onTestFileStart(test);
+    } else {
+      super.onTestStart(test);
+    }
+  }
+
+  onTestFileResult(test, testResult, aggregatedResult) {
+    this._pendingTestFiles.delete(test.path);
+
+    // @ts-ignore Precaution in case Jest migrates to the new signature
+    if (typeof super.onTestFileResult === 'function') {
+      // @ts-ignore
+      super.onTestFileResult(test, testResult, aggregatedResult);
+    } else {
+      super.onTestResult(test, testResult, aggregatedResult);
+    }
+
+    if (this._lastRunComplete && this._pendingTestFiles.size === 0) {
+      this._lastRunComplete.resolve(aggregatedResult);
+    }
+  }
+
   /**
-   * @param {Set<import('@jest/reporters').TestContext>} contexts
-   * @param {import('@jest/reporters').AggregatedResult} results
+   * @param {Set<import('@jest/reporters').TestContext>} testContexts
+   * @param {import('@jest/reporters').AggregatedResult} aggregatedResult
    * @returns {Promise<any> | null}
    */
-  // @ts-ignore
-  onRunComplete(contexts, results) {
-    // Both `_lastResults` and `_runCompletePromise` are used to prevent
-    // a bug in Jest, where `onRunComplete` is called multiple times when
-    // Jest runs with `--bail` and multiple workers, and `onRunComplete`
-    // is implemented as an asynchronous long-running operation.
-    this._lastResults = [contexts, results];
+  // @ts-ignore We need to use the complete signature, not the one from the base class
+  onRunComplete(testContexts, aggregatedResult) {
     if (!this._runCompletePromise) {
-      // @ts-expect-error TS2554: Expected 0 arguments, but got 2.
-      super.onRunComplete(...this._lastResults);
-      this._runCompletePromise = this._onRunComplete();
+      // Both `_lastRunComplete` and `_runCompletePromise` are used to prevent
+      // a bug in Jest, where `onRunComplete` is called multiple times when
+      // Jest runs with `--bail` and multiple workers, and `onRunComplete`
+      // is implemented as an asynchronous long-running operation.
+      this._lastRunComplete = this._pendingTestFiles.size === 0
```

Comment:
OK so as a name, `_lastRunComplete` is kind of confusing. It needs a name that better conveys the meaning behind it, namely of "aggregated results of the last call to onRunComplete()" - I reckon? WDYT?

Reply:
In any case, the flow here if incredibly hard to follow. Jest's fault, I know - but we need this thing somehow simplified. On the one hand, `onRunComplete` is triggered multiple times, awaiting the currently set `_lastRunComplete`. On the other, `onTestFileResult` only resolves the last `_lastRunComplete` (lol). What flow is being mitigated here, really?... I saw some of that is described in the issue's main description, but ideally, the code itself should convey the complete picture, regardless (or at least with complimentary commenting work).